### DISCUSSION TITLE
one-to-many relation beteen records from the same collection

### DISCUSSION THREAD

**Discussion by shot-codes** [pocketbase/pocketbase#1261] [Q&A] [answered]
I have a recursive data structure that I am trying to find a good way to model in pocketbase. Is it at all possible to have a one-to-many relation between records all within the same collection? Looking forward to hearing ideas on how to best tackle this!

For reference, this is what my data structure looks like. The field in question being `children`.
```ts
export interface Planet {
  text: string;
  radius: number;
  material: Material;
  children: Planet[];
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
> Is it at all possible to have a one-to-many relation between records all within the same collection?

Yes, you can create a multi `relation` field pointing to the same collection.

But currently there are 2 limitations to the multi-value fields:
- there is no multi-match filter expression (I'm currently working on this), so for example at the moment you can't have a filter that will require all `children` items to meet a certain condition, eg. `children.radius > 30` will result to `true` if there is atleast 1 Planet with `radius > 30`.
- you have to assign the full relation ids array every time when removing or adding new child (in the near future there will be support for "field modifiers", aka. something like `"children+": newId`)

Another more traditional SQL approach would be to create a second junction collection that could contain 2 `relation` fields representing the child-parent relationship.

**Reply by shot-codes** [reply]
I'm happy to hear this, the limitations are not so relevant for my use case. I realize now I wasn't seeing the option when creating the collection because it didn't exist yet for me to select it. Works as expected after creating the collection before trying to define the field.

Thanks for the response, so far really like what pocketbase is offering, great work. I think the name is very well suited, feels like I keep a whole backend in my pocket :)

Oh, one further question. Will I be able to make use of the relation expansion to expand all nested instances of Planet in children, and avoid having to make more than a single request to pb?

**Reply by ganigeorgiev** [reply]
> Oh, one further question. Will I be able to make use of the relation expansion to expand all nested instances of Planet in children, and avoid having to make more than a single request to pb?


Yes, they will be expanded as array or records, but it is not recursive automatically, aka. you'll have to specify the depth level, eg. `?expand=children.children` will produce a json like:

```js
{
  "text": "...",
  "radius": 10,
  "material": "...",
  "children": ["id1", "id2", "id3"],
  "expand": {
    "children": [
      {
        "text": "...",
        "radius": 10,
        "material": "...",
        "children": ["subid1", "subid2", "subid3"],
        "expand": {
          "children": [...]
        }
      },
      ...
    ]
  }
}
```

**Reply by shot-codes** [reply]
Okay awesome! How does it handle if ask for expansion in a level that doesn't exist for a particular record. I.e, Can I query with `?expand=children.children.children.children....children` and it will gracefully handle records with a lower nesting level?

**Reply by ganigeorgiev** [reply]
It should work just fine (aka. it will try to expand as long as there are relations), but please note that it is limited to max 6 level depth.
If you haven't already, you can check https://pocketbase.io/docs/expanding-relations/.

**Comment by renatocaliari** [comment]
> * you have to assign the full relation ids array every time when removing or adding new child (in the near future there will be support for "field modifiers", aka. something like `"children+": newId`)

@ganigeorgiev, is there any "field modifier" currently available (to add/remove) or is the alternative still to pass the full list of items of a field when creating or updating a record?

**Reply by ganigeorgiev** [reply]
They are not documented very well at the moment, but the field modifiers are available with PocketBase v0.11+ (_the fields that supports them are listed in the [Overview fields table](https://pocketbase.io/docs/collections/)_).

**Reply by renatocaliari** [reply]
Wow, that's great to know! Thank you for your prompt response @ganigeorgiev. 

I have gone through the documentation and I'm unsure about how to use the modifiers with the js-sdk. Could you please guide me on how to use them?

**Reply by ganigeorgiev** [reply]
Here are couple examples:

```js
// assuming that the original value of someRelField is ["id1", "id2", "id3"]
await pb.collection("example").update("RECORD_ID", {
    // append single id (results in ["id1", "id2", "id3", "id4"])
    "someRelField+": "id4", 

    // append multiple ids (results in ["id1", "id2", "id3", "id4", "id5"])
    "someRelField+": ["id4", "id5"],

    // remove single id (results in ["id2", "id3"])
    "someRelField-": "id1",
    
    // remove multiple ids (results in ["id2"])
    "someRelField-": ["id1", "id3"],
})
```

_Note that it is allowed to combine `+`, `-` and even directly assigning an array value in the same request body but the order of the operations are not guaranteed to be exactly as in the submitted body since Go maps are unordered/doesn't preserve the insertion order._

**Reply by renatocaliari** [reply]
Thank you! 
It was my mistake for not testing the exact structure of your quote that I had copied from your previous message, `"children+": newId`. 

But now, with your examples, I understood even better
